### PR TITLE
Hyscale exec command

### PR DESCRIPTION
**Pull request by satya-k8s**
exec command using kubectl

### CODE REVIEW

File: commons/src/main/java/io/hyscale/commons/commands/CommandExecutor.java [hunk 1]
```diff
@@ -200,4 +209,18 @@ private static OutputStream getOutputStream(File outputFile) throws FileNotFound
         return new StringOutputStream();
     }
     
+    
+	public static void executeExec(List<String> commands) {
```
Comment (tushar-tushar) [inline commons/src/main/java/io/hyscale/commons/commands/CommandExecutor.java]: Instead of eating exception we need to throw HyscaleException.
For commands should we have the argument as List or String?
Add null checks where required.
Also should we return exit code to know if command was successful or not?
Also what were the values of exit code if someone stops the command?

File: deployer/deployer-services/src/main/java/io/hyscale/deployer/services/deployer/Deployer.java [hunk 2]
```diff
@@ -108,6 +108,11 @@ public default void waitForDeployment(DeploymentContext context, ProgressHandler
 	public List<ReplicaInfo> getReplicas(AuthConfig authConfig, String appName, String serviceName, String namespace, 
 	        boolean isFilter) throws HyscaleException;
 	
+	
+	
+	
+	 public void exec(K8sAuthorisation k8sAuthorisation, String servicename, String appname, String namespace, String replica);
```
Comment (tushar-tushar) [inline deployer/deployer-services/src/main/java/io/hyscale/deployer/services/deployer/Deployer.java]: Add Throws HyscaleException

File: deployer/deployer-services/src/main/java/io/hyscale/deployer/services/deployer/impl/KubernetesDeployer.java [hunk 3]
```diff
@@ -169,6 +192,36 @@ public boolean authenticate(AuthConfig authConfig) {
...
+			K8sConfigFileAuth k8sConfigFileAuth = (K8sConfigFileAuth) k8sAuthorisation;
+			kubeConfig = getKubeConfig(k8sConfigFileAuth);
+			if (kubeConfig.getAccessToken() != null) {
+				commands = ExecCommandUtil.getExecCommand(replicaName, namespace, kubeConfig.getAccessToken());
+				CommandExecutor.executeExec(commands);
+			}
+		}
+	}
+
+	private KubeConfig getKubeConfig(K8sConfigFileAuth k8sConfigFileAuth) {
```
Comment (tushar-tushar) [inline deployer/deployer-services/src/main/java/io/hyscale/deployer/services/deployer/impl/KubernetesDeployer.java]: As discussed this should not be required right? Directly passing the file to the command works?

File: deployer/deployer-services/src/main/java/io/hyscale/deployer/services/deployer/impl/KubernetesDeployer.java [hunk 4]
```diff
@@ -169,6 +192,36 @@ public boolean authenticate(AuthConfig authConfig) {
...
+    
+	@Override
+	public void exec(K8sAuthorisation k8sAuthorisation, String servicename, String appname, String namespace,
+			String replicaName) {
+
+		List<String> commands = null;
+		KubeConfig kubeConfig = null;
+		if (k8sAuthorisation != null && k8sAuthorisation.getK8sAuthType() == K8sAuthType.BASIC_AUTH) {
+			K8sBasicAuth k8sBasicAuth = (K8sBasicAuth) k8sAuthorisation;
+			commands = ExecCommandUtil.getExecCommand(replicaName, namespace, k8sBasicAuth);
```
Comment (tushar-tushar) [inline deployer/deployer-services/src/main/java/io/hyscale/deployer/services/deployer/impl/KubernetesDeployer.java]: This command is not executed later on

File: deployer/deployer-services/src/main/java/io/hyscale/deployer/services/deployer/impl/KubernetesDeployer.java [hunk 5]
```diff
@@ -169,6 +192,36 @@ public boolean authenticate(AuthConfig authConfig) {
...
+		} else if (k8sAuthorisation != null && k8sAuthorisation.getK8sAuthType() == K8sAuthType.KUBE_CONFIG_FILE) {
+			K8sConfigFileAuth k8sConfigFileAuth = (K8sConfigFileAuth) k8sAuthorisation;
+			kubeConfig = getKubeConfig(k8sConfigFileAuth);
+			if (kubeConfig.getAccessToken() != null) {
+				commands = ExecCommandUtil.getExecCommand(replicaName, namespace, kubeConfig.getAccessToken());
+				CommandExecutor.executeExec(commands);
+			}
+		}
+	}
```
Comment (tushar-tushar) [inline deployer/deployer-services/src/main/java/io/hyscale/deployer/services/deployer/impl/KubernetesDeployer.java]: Extract out the null check and throw exception if k8sAuthorisation is null
Then should we use switch with k8sAuthorisation.getK8sAuthType()? As later we may have more cases to support.
Should we have the logic to get commands in DeployCommandProvider on the lines of ImageCommandProvider
It would take k8sAuthorisation and other variable and return the command. No need to have that logic here.

File: deployer/deployer-services/src/main/java/io/hyscale/deployer/services/util/ExecCommandUtil.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+	public static final String TOKEN="--token";
+	
+	public static List<String> getExecCommand(String replicaName, String namespace, String accessToken){
+		List<String> commands=getCommand(replicaName, namespace);
+        commands.add(TOKEN);
+        commands.add(accessToken);
+        return commands;
+	}
+	
+	public static List<String> getExecCommand(String replicaName, String namespace,K8sBasicAuth k8sBasicAuth ){
```
Comment (tushar-tushar) [inline deployer/deployer-services/src/main/java/io/hyscale/deployer/services/util/ExecCommandUtil.java]: This can have the logic as specified in the deployer.
Make the methods non static

File: workflow-controller/controller-service/src/main/java/io/hyscale/controller/commands/HyscaleExecCommand.java [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+	 
+	 @Autowired
+	 private Deployer deployer;
+
+	@Override
+	public Integer call() throws Exception {
+		if (!CommandUtil.isInputValid(this)) {
+			return ToolConstants.INVALID_INPUT_ERROR_CODE;
+		}
+		authConfigBuilder.getAuthConfig();
```
Comment (tushar-tushar) [inline workflow-controller/controller-service/src/main/java/io/hyscale/controller/commands/HyscaleExecCommand.java]: This will give the K8sAuthConfig, that can be passed to the deployer method.
No need to read the file manually

File: workflow-controller/controller-service/src/main/java/io/hyscale/controller/commands/HyscaleExecCommand.java [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+		K8sConfigFileAuth k8sConfigFileAuth=new K8sConfigFileAuth();
+		k8sConfigFileAuth.setK8sConfigFile(new File(System.getProperty("user.home") + "/.kube/config"));
+		K8sAuthorisation k8sAuthorisation=k8sConfigFileAuth;
+		deployer.exec(k8sAuthorisation,serviceName, appName, namespace, replicaName);
+		return null;
```
Comment (tushar-tushar) [inline workflow-controller/controller-service/src/main/java/io/hyscale/controller/commands/HyscaleExecCommand.java]: This return is used as exit code.
On success return 0.
Else deployer should throw the exception. catch it, log it and rethrow

File: commons/src/main/java/io/hyscale/commons/commands/CommandExecutor.java (hunk 1 above)
Comment (SameeraGrandhi) [reply commons/src/main/java/io/hyscale/commons/commands/CommandExecutor.java]: method name can be better like executeWithParentIO , should be relevant to class contract

File: deployer/deployer-services/src/main/java/io/hyscale/deployer/services/util/ExecCommandUtil.java [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+import io.hyscale.commons.models.K8sBasicAuth;
+
+public class ExecCommandUtil {
+	public static final String KUBECTL ="kubectl";
+	public static final String EXEC="exec";
+	public static final String IT="-it";
+	public static final String NAMESPACE="-n";
+	public static final String BASH="bash";
+	public static final String TOKEN="--token";
+
```
Comment (SameeraGrandhi) [inline deployer/deployer-services/src/main/java/io/hyscale/deployer/services/util/ExecCommandUtil.java]: How do you manage kubeconfig with --kubeconfig option

File: deployer/deployer-services/src/main/java/io/hyscale/deployer/services/util/ExecCommandUtil.java [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+	}
+	
+	public static List<String> getExecCommand(String replicaName, String namespace,K8sBasicAuth k8sBasicAuth ){
+		List<String> commands=getCommand(replicaName, namespace);
+        return commands;
+	}
+	
+	private static List<String> getCommand(String replicaName, String namespace) {
+		List<String> commands = new ArrayList<>();
+		commands.add(KUBECTL);
```
Comment (SameeraGrandhi) [inline deployer/deployer-services/src/main/java/io/hyscale/deployer/services/util/ExecCommandUtil.java]: As specified use --kubeconfig option also

File: commons/src/main/java/io/hyscale/commons/commands/CommandExecutor.java [hunk 11]
```diff
@@ -200,4 +208,22 @@ private static OutputStream getOutputStream(File outputFile) throws FileNotFound
...
     }
     
+    
+	public static int executeWithParentIO(List<String> commands){
+			ProcessBuilder processBuilder = new ProcessBuilder(commands);
+			processBuilder.redirectOutput(ProcessBuilder.Redirect.INHERIT);
+			processBuilder.redirectInput(ProcessBuilder.Redirect.INHERIT);
+			processBuilder.redirectErrorStream(true);
+			Process process;
+			int exitCode=0;
```
Comment (tushar-tushar) [inline commons/src/main/java/io/hyscale/commons/commands/CommandExecutor.java]: Keep default value as 1

File: workflow-controller/controller-service/src/main/java/io/hyscale/controller/commands/HyscaleExecCommand.java [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+	private Deployer deployer;
+	 @Autowired
+	 private K8sAuthConfigBuilder authConfigBuilder;
+
+	@Override
+	public Integer call() throws Exception {
+		if (!CommandUtil.isInputValid(this)) {
+			return ToolConstants.INVALID_INPUT_ERROR_CODE;
+		}
+		K8sConfigFileAuth k8sConfigFileAuth=(K8sConfigFileAuth) authConfigBuilder.getAuthConfig();
```
Comment (tushar-tushar) [inline workflow-controller/controller-service/src/main/java/io/hyscale/controller/commands/HyscaleExecCommand.java]: Do we need type casting here? Can't we directly send the parameter?

File: deployer/deployer-services/src/main/java/io/hyscale/deployer/services/deployer/Deployer.java [hunk 13]
```diff
@@ -108,6 +108,20 @@ public default void waitForDeployment(DeploymentContext context, ProgressHandler
...
+	 * Exec into  Pods
+	 * @param k8sAuthorisation
+	 * @param servicename
+	 * @param appname
+	 * @param namespace
+	 * @param replica
+	 * @return exitCode
+	 * @throws HyscaleException
+	 */
+	 public int exec(K8sAuthorisation k8sAuthorisation, String servicename, String appname, String namespace, String replica) throws HyscaleException;
```
Comment (tushar-tushar) [inline deployer/deployer-services/src/main/java/io/hyscale/deployer/services/deployer/Deployer.java]: Would we need the return type in case of platform?
Also camel case naming convention
Also name the variable order: K8sAuth, appName, ServiceName, replicaName, namespace

File: deployer/deployer-services/src/main/java/io/hyscale/deployer/services/deployer/impl/KubernetesDeployer.java [hunk 14]
```diff
@@ -169,6 +193,55 @@ public boolean authenticate(AuthConfig authConfig) {
...
+			kubeConfig = getKubeConfig(k8sConfigFileAuth);
+			if (kubeConfig.getAccessToken() != null) {
+				commands = deployCommandProvider.getExecCommandByAccessToken(replicaName, namespace,
+						kubeConfig.getAccessToken());
+				exitCode = CommandExecutor.executeWithParentIO(commands);
+			}
+			break;
+
+		default:
+			break;
```
Comment (tushar-tushar) [inline deployer/deployer-services/src/main/java/io/hyscale/deployer/services/deployer/impl/KubernetesDeployer.java]: In default case should we throw exception saying K8SAuthorisation not yet supported?

File: deployer/deployer-services/src/main/java/io/hyscale/deployer/services/deployer/impl/KubernetesDeployer.java [hunk 15]
```diff
@@ -169,6 +193,55 @@ public boolean authenticate(AuthConfig authConfig) {
...
+	@Override
+	public int exec(K8sAuthorisation k8sAuthorisation, String servicename, String appname, String namespace,
+			String replicaName) throws HyscaleException {
+
+		if (k8sAuthorisation == null) {
+			throw new HyscaleException(DeployerErrorCodes.K8SAUTHORISATION_NOT_FOUND);
+		}
+		List<String> commands = null;
+		KubeConfig kubeConfig = null;
+		int exitCode = 0;
```
Comment (tushar-tushar) [inline deployer/deployer-services/src/main/java/io/hyscale/deployer/services/deployer/impl/KubernetesDeployer.java]: keep default value as 1

File: deployer/deployer-services/src/main/java/io/hyscale/deployer/services/deployer/impl/KubernetesDeployer.java [hunk 16]
```diff
@@ -169,6 +193,55 @@ public boolean authenticate(AuthConfig authConfig) {
...
+
+		switch (k8sAuthorisation.getK8sAuthType()) {
+
+		case BASIC_AUTH:
+			K8sBasicAuth k8sBasicAuth = (K8sBasicAuth) k8sAuthorisation;
+			commands = deployCommandProvider.getExecCommandByBasicAuth(replicaName, namespace, k8sBasicAuth);
```
Comment (tushar-tushar) [inline deployer/deployer-services/src/main/java/io/hyscale/deployer/services/deployer/impl/KubernetesDeployer.java]: This command is not executed?

File: deployer/deployer-services/src/main/java/io/hyscale/deployer/services/deployer/impl/KubernetesDeployer.java [hunk 17]
```diff
@@ -169,6 +193,55 @@ public boolean authenticate(AuthConfig authConfig) {
...
+			break;
+
+		case KUBE_CONFIG_FILE:
+			K8sConfigFileAuth k8sConfigFileAuth = (K8sConfigFileAuth) k8sAuthorisation;
+			kubeConfig = getKubeConfig(k8sConfigFileAuth);
+			if (kubeConfig.getAccessToken() != null) {
+				commands = deployCommandProvider.getExecCommandByAccessToken(replicaName, namespace,
```
Comment (tushar-tushar) [inline deployer/deployer-services/src/main/java/io/hyscale/deployer/services/deployer/impl/KubernetesDeployer.java]: Instead of this can we use getExecCommandByKubeconfig and pass the file path?

File: commons/src/main/java/io/hyscale/commons/commands/provider/DeployCommandProvider.java [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+	}
+	
+	public List<String> getExecCommandByKubeconfig(String replicaName, String namespace, String kubeConfig) {
+		List<String> commands = getCommand(replicaName, namespace);
+		commands.add(KUBECONFIG);
+		commands.add(kubeConfig);
+		return commands;
+	}
+	
+	public List<String> getExecCommandByBasicAuth(String replicaName, String namespace, K8sBasicAuth k8sBasicAuth) {
```
Comment (tushar-tushar) [inline commons/src/main/java/io/hyscale/commons/commands/provider/DeployCommandProvider.java]: This method is not using k8sBasicAuth at all?

File: commons/src/main/java/io/hyscale/commons/commands/CommandExecutor.java [hunk 19]
```diff
@@ -222,7 +222,7 @@ public static int executeWithParentIO(List<String> commands){
 			} catch (Exception e) {
 				 HyscaleException ex = new HyscaleException(e, CommonErrorCode.FAILED_TO_EXECUTE_COMMAND);
 		         logger.error("Error while exec  into pod with error code {}, error {}", exitCode, ex);
-		         return exitCode;
+		         return ex.getCode();
```
Comment (tushar-tushar) [inline commons/src/main/java/io/hyscale/commons/commands/CommandExecutor.java]: What would be the value of this? Since it is not set would it be null?

File: commons/src/main/java/io/hyscale/commons/commands/provider/DeployCommandProvider.java [hunk 20]
```diff
@@ -30,24 +31,19 @@
...
-		commands.add(accessToken);
-		return commands;
-	}
-	
-	public List<String> getExecCommandByKubeconfig(String replicaName, String namespace, String kubeConfig) {
+	public List<String> getExecCommandByKubeConfig(String replicaName, String namespace, File kubeConfigFile) {
 		List<String> commands = getCommand(replicaName, namespace);
 		commands.add(KUBECONFIG);
-		commands.add(kubeConfig);
+		commands.add(kubeConfigFile.toString());
```
Comment (tushar-tushar) [inline commons/src/main/java/io/hyscale/commons/commands/provider/DeployCommandProvider.java]: Why toString() and not getPath or getAbsolutePath()?

File: commons/src/main/java/io/hyscale/commons/commands/provider/DeployCommandProvider.java [hunk 21]
```diff
@@ -30,24 +31,19 @@
...
 		return commands;
 	}
 	
 	public List<String> getExecCommandByBasicAuth(String replicaName, String namespace, K8sBasicAuth k8sBasicAuth) {
 		List<String> commands = getCommand(replicaName, namespace);
+		commands.add(TOKEN);
+		commands.add(k8sBasicAuth.getToken());
```
Comment (tushar-tushar) [inline commons/src/main/java/io/hyscale/commons/commands/provider/DeployCommandProvider.java]: Null check?

File: deployer/deployer-services/src/main/java/io/hyscale/deployer/services/deployer/impl/KubernetesDeployer.java [hunk 22]
```diff
@@ -196,51 +192,56 @@ public boolean authenticate(AuthConfig authConfig) {
...
-		KubeConfig kubeConfig = null;
-		int exitCode = 0;
+		int exitCode = 1;
 
 		switch (k8sAuthorisation.getK8sAuthType()) {
 
 		case BASIC_AUTH:
 			K8sBasicAuth k8sBasicAuth = (K8sBasicAuth) k8sAuthorisation;
 			commands = deployCommandProvider.getExecCommandByBasicAuth(replicaName, namespace, k8sBasicAuth);
+			exitCode = CommandExecutor.executeWithParentIO(commands);
```
Comment (tushar-tushar) [inline deployer/deployer-services/src/main/java/io/hyscale/deployer/services/deployer/impl/KubernetesDeployer.java]: We can keep this outside of switch as this is common

File: deployer/deployer-services/src/main/java/io/hyscale/deployer/services/deployer/impl/KubernetesDeployer.java [hunk 23]
```diff
@@ -196,51 +192,56 @@ public boolean authenticate(AuthConfig authConfig) {
...
-	private KubeConfig getKubeConfig(K8sConfigFileAuth k8sConfigFileAuth) {
-		InputStreamReader isr = null;
-		try {
-			isr = new FileReader(k8sConfigFileAuth.getK8sConfigFile());
-		} catch (FileNotFoundException e) {
-			logger.error("Error while reading kubeconfig file {}", e.toString());
+	
+	private boolean validateReplica(AuthConfig authConfig, String appName,String serviceName, String namespace, String replicaName) throws HyscaleException {
+		List<ReplicaInfo> replicaInfos=null;
+		replicaInfos=getReplicas(authConfig, appName, serviceName, namespace, false);
```
Comment (tushar-tushar) [inline deployer/deployer-services/src/main/java/io/hyscale/deployer/services/deployer/impl/KubernetesDeployer.java]: This can be in single line

File: deployer/deployer-services/src/main/java/io/hyscale/deployer/services/deployer/impl/KubernetesDeployer.java [hunk 24]
```diff
@@ -196,51 +192,56 @@ public boolean authenticate(AuthConfig authConfig) {
...
+		if(replicaInfos==null && replicaInfos.isEmpty()) {
```
Comment (tushar-tushar) [inline deployer/deployer-services/src/main/java/io/hyscale/deployer/services/deployer/impl/KubernetesDeployer.java]: null || empty

File: deployer/deployer-services/src/main/java/io/hyscale/deployer/services/deployer/impl/KubernetesDeployer.java [hunk 25]
```diff
@@ -196,51 +192,56 @@ public boolean authenticate(AuthConfig authConfig) {
     
     
 	@Override
-	public int exec(K8sAuthorisation k8sAuthorisation, String servicename, String appname, String namespace,
+	public void exec(K8sAuthorisation k8sAuthorisation, String serviceName, String appName, String namespace,
 			String replicaName) throws HyscaleException {
-
+		boolean validateReplica=validateReplica(k8sAuthorisation, appName, serviceName, namespace, replicaName);
```
Comment (tushar-tushar) [inline deployer/deployer-services/src/main/java/io/hyscale/deployer/services/deployer/impl/KubernetesDeployer.java]: rename variable to isReplicaValid

File: deployer/deployer-services/src/main/java/io/hyscale/deployer/services/deployer/impl/KubernetesDeployer.java [hunk 26]
```diff
@@ -196,51 +192,56 @@ public boolean authenticate(AuthConfig authConfig) {
...
+		replicaInfos=getReplicas(authConfig, appName, serviceName, namespace, false);
+		if(replicaInfos==null && replicaInfos.isEmpty()) {
+		 throw new HyscaleException(DeployerErrorCodes.SERVICE_NOT_FOUND, serviceName,appName);
 		}
-		return KubeConfig.loadKubeConfig(isr);
+		Set<String> replicaNames=new HashSet<String>();
+		for(ReplicaInfo replicaInfo:replicaInfos) {
+			replicaNames.add(replicaInfo.getName());
+		}
+		return replicaNames.contains(replicaName);
```
Comment (tushar-tushar) [inline deployer/deployer-services/src/main/java/io/hyscale/deployer/services/deployer/impl/KubernetesDeployer.java]: Why is this required? Can we directly use stream instead, something like: replicaInfos.stream().anyMatch(existing -> existing.getName().equals(replicaName));

File: workflow-controller/controller-service/src/main/java/io/hyscale/controller/commands/HyscaleExecCommand.java [hunk 27]
```diff
@@ -61,8 +61,7 @@ public Integer call() throws Exception {
 		if (!CommandUtil.isInputValid(this)) {
 			return ToolConstants.INVALID_INPUT_ERROR_CODE;
 		}
-		K8sConfigFileAuth k8sConfigFileAuth=(K8sConfigFileAuth) authConfigBuilder.getAuthConfig();
-		return deployer.exec(k8sConfigFileAuth,serviceName, appName, namespace, replicaName);
+	    deployer.exec((K8sAuthorisation)authConfigBuilder.getAuthConfig(),serviceName, appName, namespace, replicaName);
+	    return null;
```
Comment (tushar-tushar) [inline workflow-controller/controller-service/src/main/java/io/hyscale/controller/commands/HyscaleExecCommand.java]: This should return 0
